passes/cmds/sta.cc

```diff
@@ -0,0 +1,295 @@
+						auto r = endpoints.insert(bit);
+						if (r.second || r.first->second.required < it->second.first) {
+							r.first->second.sink = cell;
+							r.first->second.port = conn.first;
+							r.first->second.required = it->second.first;
+						}
+					}
+					if (cell->output(conn.first)) {
+						dst_bits.insert(std::make_pair(bit,namebit));
+						auto &d = data[bit];
+						d.driver = cell;
+						d.dst_port = conn.first;
+
+						auto it = t.arrival.find(namebit);
+						if (it == t.arrival.end())
+							continue;
+						const auto &s = it->second.second;
+						if (cell->hasPort(s.name)) {
+							auto s_bit = sigmap(cell->getPort(s.name)[s.offset]);
+							if (s_bit.wire)
+								data[s_bit].fanouts.emplace_back(bit,it->second.first,s.name);
+						}
+					}
+				}
+			}
+
+			for (const auto &s : src_bits)
+				for (const auto &d : dst_bits) {
+					auto it = t.comb.find(TimingInfo::BitBit(s.second,d.second));
+					if (it == t.comb.end())
+						continue;
+					data[s.first].fanouts.emplace_back(d.first,it->second,s.second.name);
+				}
+		}
+
+		for (auto port_name : module->ports) {
+			auto wire = module->wire(port_name);
+			if (wire->port_input) {
+				for (const auto &b : sigmap(wire))
+					queue.emplace_back(b);
+				// All primary inputs to arrive at time zero
+				wire->set_intvec_attribute(ID(sta_arrival), std::vector<int>(GetSize(wire), 0));
+			}
+			if (wire->port_output)
+				for (const auto &b : sigmap(wire))
+					if (b.wire)
+						endpoints.insert(b);
+		}
+	}
+
+	void run()
+	{
+		while (!queue.empty()) {
+			auto b = queue.front();
+			queue.pop_front();
+			auto it = data.find(b);
+			if (it == data.end())
+				continue;
+			const auto& src_arrivals = b.wire->get_intvec_attribute(ID(sta_arrival));
+			log_assert(GetSize(src_arrivals) == GetSize(b.wire));
+			auto src_arrival = src_arrivals[b.offset];
+			for (const auto &d : it->second.fanouts) {
+				const auto &dst_bit = std::get<0>(d);
+				auto dst_arrivals = dst_bit.wire->get_intvec_attribute(ID(sta_arrival));
+				if (dst_arrivals.empty())
+					dst_arrivals = std::vector<int>(GetSize(dst_bit.wire), -1);
+				else
+					log_assert(GetSize(dst_arrivals) == GetSize(dst_bit.wire));
+				auto &dst_arrival = dst_arrivals[dst_bit.offset];
+				auto new_arrival = src_arrival + std::get<1>(d);
+				if (dst_arrival < new_arrival) {
+					dst_arrival = std::max(dst_arrival, new_arrival);
+					dst_bit.wire->set_intvec_attribute(ID(sta_arrival), dst_arrivals);
+					queue.emplace_back(dst_bit);
+
+					data[dst_bit].backtrack = b;
+					data[dst_bit].src_port = std::get<2>(d);
+				}
+				auto it = endpoints.find(dst_bit);
+				if (it != endpoints.end()) {
+					new_arrival += it->second.required;
+					if (new_arrival > maxarrival) {
+						maxarrival = new_arrival;
+						maxbit = dst_bit;
+					}
+				}
+			}
+		}
+
+		log("Latest arrival time in '%s' is %d:\n", log_id(module), maxarrival);
+		auto b = maxbit;
+		const auto &e = endpoints.at(maxbit);
```

Comment:
I've been playing around with adding Gowin specify blocks, and while working on it keep hitting aborts when `maxbit` is not in `endpoints`. Could you add some better error handling for that?

Reply:
Just to check we're on the same page first: would you agree that that if `maxbit` is not in `endpoints`, that means the latest arrival time does not terminate at a primary output or sequential input, and so handling it would mean just emitting an error message? Or am I missing something?

Reply:
Yeah, an error message would be much more graceful than a `std::out_of_range` exception.

Though I feel I should ask what the definition of sequential input is here; a `posedge` specify?

Additionally, while discussing this pass, I get warnings on not having timing arcs for `VCC`/`GND`, which as the name implies are just constant drivers; as far as I can tell this isn't actually something you can communicate to `sta`.

Reply:
> Yeah, an error message would be much more graceful than a `std::out_of_range` exception.

Thanks for pointing that out. I admit it is not very friendly to any one but me right now. Resolved in 5b26d4b64c8586c29e5e8e65ead44ed0af945282.
 
> Though I feel I should ask what the definition of sequential input is here; a `posedge` specify?

Endpoints are cell input ports with "required" (`$setup`) time constraints: https://github.com/YosysHQ/yosys/blob/5b26d4b64c8586c29e5e8e65ead44ed0af945282/passes/cmds/sta.cc#L94-L97

Or are primary outputs:
https://github.com/YosysHQ/yosys/blob/5b26d4b64c8586c29e5e8e65ead44ed0af945282/passes/cmds/sta.cc#L140-L143


> 
> Additionally, while discussing this pass, I get warnings on not having timing arcs for `VCC`/`GND`, which as the name implies are just constant drivers; as far as I can tell this isn't actually something you can communicate to `sta`.

Reply:
> Can you think of a system to identify these calls *without* having to tag them with an attribute to suppress the no-timing-arcs warning?

The most obvious property that comes to mind is that at the very least Gowin `VCC`/`GND` have no input ports (I can arrange this for `synth_intel_alm` too, as the `__MISTRAL_{VCC,GND}` cells are also like this). Lacking any inputs means they must surely be constant at runtime, unless there's some Verilog trickery that I'm not aware of.

Reply:
Alternatively you could check if a cell has only constant inputs, but for a simple test, the "no inputs" method seems like a good rule of thumb.

Reply:
@ZirconiumX The "no timing arcs" warning message should now be suppressed by ee52e3a. Is it what you was looking for?

Reply:
Yep!